File: 021.rb
# Let d(n) be defined as the sum of proper divisors of n 
# (numbers less than n which divide evenly into n).
# If d(a) = b and d(b) = a, where a ≠ b, then a and b are an 
# amicable pair and each of a and b are called amicable numbers.
# Evaluate the sum of all the amicable numbers under 10000.

def sum_proper_divisors(n)
    (1..(n/2 + 1)).find_all { |x| n%x == 0 }.reduce(:+)
end

def amicable?(n)
    proper_divisors_sum = sum_proper_divisors(n)
    proper_divisors_sum != n and sum_proper_divisors(proper_divisors_sum) == n
end

puts (1...10000).find_all { |x| amicable?(x) }.reduce(:+)

File: clojure/src/project_euler_clojure/021.clj
;; Let d(n) be defined as the sum of proper divisors of n
;; (numbers less than n which divide evenly into n).
;; If d(a) = b and d(b) = a, where a ≠ b, then a and b are an
;; amicable pair and each of a and b are called amicable numbers.
;; Evaluate the sum of all the amicable numbers under 10000.

(ns project-euler-clojure.021)

(defn proper-divisors [n]
  (->> (range 1 (inc (/ n 2)))
       (filter (comp zero? (partial mod n)))))

(defn sum-proper-divisors [n]
  (->> (proper-divisors n)
       (reduce +)))

(defn amicable? [n]
  (and (not= (sum-proper-divisors n) n)
       (= (sum-proper-divisors (sum-proper-divisors n)) n)))

(defn problem-021 []
  (->> (range 1 10000)
       (filter amicable?)
       (reduce +)))
